README.md
# RepeatableQuestFilter

Overview
--------

This addon enables focused farming or leveling by filtering __repeatable quests__ received by certain __quest givers__.  At the moment, I am supporting the __Dark Brotherhood__ and __Thieves Guild__ repeatable quests obtained through the *Marked for Death* and *Tip Board*, respectively.

Usage
-------

Simply spam "E" at the Quest Giver to allow the filter to only accept the quests according to your criteria.

Filters
-------

Thieves Guild:

- Crime Spree
- Launder (The Covetous Countess)

Dark Brotherhood:

- Kill Spree
    - Specific Locations

Bugs & Requests
---------------

If you experience a bug or have any feature requests, please let me know in the [Issue Tracker](https://github.com/alexgurrola/RepeatableQuestFilter/issues) on the Github Repository.  Since this is my first addon and first time playing Elder Scrolls Online, there will likely be much more to be desired, so don't be shy.  ;)

Contributions
-------------

You are welcome to make changes and submit them for review as a [Pull Request](https://github.com/alexgurrola/RepeatableQuestFilter/pulls) in the Open Source [Repository](https://github.com/alexgurrola/RepeatableQuestFilter) on Github.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2018-05-06

### Added
- Changelog
- Donation Button

### Changed
- Nested Control Settings
- Require Reload for Setting Changes
- Allow Cutpurse Quest for Dailies

## [1.1.0] - 2018-05-03

### Changed
- New Configuration Version

### Fixed
- Settings Panel

## [1.0.0] - 2018-05-02

- Initial Release

[Unreleased]: https://github.com/alexgurrola/RepeatableQuestFilter/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/alexgurrola/RepeatableQuestFilter/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/alexgurrola/RepeatableQuestFilter/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/alexgurrola/RepeatableQuestFilter/tree/v1.0.0


RepeatableQuestData.lua
-- zones for usage
local zones = {
  [821] = true, -- Thieves Den
  [826] = true, -- Dark Brotherhood Sanctuary
}

-- localized names of the quest givers
local questGiver = {
	["Tip Board"] = true,
  ["Marked for Death"] = true,
}

-- first few characters of the quest dialogs
local dialogTG = {
	["ors that"] = false, -- Crime Spree
	["eemed th"] = false, -- The Covetous Countess
}
local dialogTGF = {
  ["The Thalmor"] = false, -- The Covetous Countess
}
local dialog = {
	["'d think"] = true,
	[" Queen h"] = true,
	["re's a b"] = true,
	["en Ayren"] = true,
	["as passi"] = true,
	["re is a "] = true,
	["annot ab"] = true,
	[" Spinner"] = true,
	["e and Jo"] = true,
	["g Aerada"] = true,
	[" don't t"] = true,
	["re's a t"] = true,
	["spouse b"] = true,
	["more Tha"] = true,
	["ry day I"] = true,
	[" of the "] = true,
	["e fool k"] = true,
	["kwasten "] = true,
	["ave a cu"] = true,
	["s one se"] = true,
	["en-ja is"] = true,
	["dbeats. "] = true,
	["ls of Ju"] = true,
	["re are s"] = true,
	["ave been"] = true,
	[" Stone O"] = true,
	[" cheerin"] = true,
	[" at peak"] = true,
	["e been d"] = true,
	["m positi"] = true,
	["py hides"] = true,
	["an't tol"] = true,
	[" being m"] = true,
	["oward hi"] = true,
	["prey has"] = true,
	["se Dorel"] = true,
	["advancem"] = true,
	["t the be"] = true,
	["se who s"] = true,
	["ealous b"] = true,
	["agitator"] = true,
	["re's an "] = true,
	["m forced"] = true,
	[" seeds o"] = true,
	["eek to g"] = true,
	["elers ma"] = true,
	["kin dish"] = true,
	[" careles"] = true,
	["lorious "] = true,
	["rect the"] = true,
	["eally ca"] = true,
	["people m"] = true,
	["lover—fo"] = true,
	["losed ar"] = true,
	["re is a "] = true,
	["rine dut"] = true,
	["n the Da"] = true,
	[" losing "] = true,
	["d slaugh"] = true,
	[" milk-dr"] = true,
	[" suspici"] = true,
}


RepeatableQuestFilter.lua
--------------------------
-- Initialize Variables --
--------------------------

RepeatableQuestFilter = {}
RepeatableQuestFilter.name = "RepeatableQuestFilter"
RepeatableQuestFilter.configVersion = 3
RepeatableQuestFilter.controls = {
    KillSpree = {
        Neutral = {},
        AldmeriDominion = {},
        DaggerfallCovenant = {},
        EbonheartPact = {},
    }
}
RepeatableQuestFilter.defaults = {
    ThievesGuild = true,
    DarkBrotherhood = true,
    CrimeSpree = true,
    Launder = false,
    KillSpree = {
        -- Gold Coast
        GoldCoast = true,
        -- Aldmeri Dominion
        Auridon = true,
        Grahtwood = true,
        Greenshade = true,
        MalabalTor = true,
        ReapersMarch = true,
        -- Daggerfall Covenant
        AlikrDesert = true,
        Bangkorai = true,
        Glenumbra = true,
        Rivenspire = true,
        Stormhaven = true,
        -- Ebonheart Pact
        Deshaan = true,
        Eastmarch = true,
        Shadowfen = true,
        Stonefalls = true,
        TheRift = true,
    }
}
RepeatableQuestFilter.filters = {}

---------------------
--  OnAddOnLoaded  --
---------------------

function OnAddOnLoaded(event, addonName)
    if addonName ~= RepeatableQuestFilter.name then
        return
    end
    RepeatableQuestFilter:Initialize(RepeatableQuestFilter)
end

--------------------------
--  Initialize Function --
--------------------------

function RepeatableQuestFilter:Initialize(self)
    self.saveData = ZO_SavedVars:New(self.name .. "Data", self.configVersion, nil, self.defaults)
    self:RepairSaveData(self)
    self:CreateSettingsWindow(self)
    self:BuildFilters(self)
    self:OverwritePopulateChatterOption(self, GAMEPAD_INTERACTION)
    self:OverwritePopulateChatterOption(self, INTERACTION) -- keyboard
    EVENT_MANAGER:UnregisterForEvent(self.name, EVENT_ADD_ON_LOADED)
end

---------------
-- Libraries --
---------------

local LAM2 = LibStub("LibAddonMenu-2.0")

--------------------
-- Internal Tools --
--------------------

-- allow debugging based on changes
RepeatableQuestFilter.debugLog = {}
function RepeatableQuestFilter:Debug(self, key, output)
    if output ~= self.debugLog[key] then
        self.debugLog[key] = output
        d(self.name .. "." .. key .. ":", output)
    end
end

function RepeatableQuestFilter:RepairSaveData(self)
    for key, value in pairs(self.defaults) do
        if (self.saveData[key] == nil) then
            self.saveData[key] = value
        end
    end
    --[[
    for key, value in pairs(self.defaults.KillSpree) do
        if (self.saveData.KillSpree[key] == nil) then
            self.saveData.KillSpree[key] = value
        end
    end
    ]]
end

------------------
-- System Hooks --
------------------

local lastInteractableName
ZO_PreHook(FISHING_MANAGER, "StartInteraction", function()
    local _, name = GetGameCameraInteractableActionInfo()
    lastInteractableName = name
end)

----------
-- Data --
----------

function RepeatableQuestFilter:BuildFilters(self)
    -- zones for usage
    self.filters.zones = {
        [821] = self.saveData.ThievesGuild, -- Thieves Den
        [826] = self.saveData.DarkBrotherhood, -- Dark Brotherhood Sanctuary
    }

    -- localized names of the quest givers
    self.filters.questGiver = {
        ["Tip Board"] = self.saveData.ThievesGuild, -- Thieves Den
        ["Marked for Death"] = self.saveData.DarkBrotherhood, -- Dark Brotherhood Sanctuary
    }

    -- first few characters of the quest dialogs
    self.filters.dialog = {
        -- Tip Board
        ["Some new fa"] = self.saveData.ThievesGuild, -- The Cutpurse's Craft (Required to get Dailies)
        ["Rumors that"] = self.saveData.CrimeSpree, -- Crime Spree
        ["Esteemed th"] = self.saveData.Launder, -- The Covetous Countess
        -- Marked for Death
        ["Demand for "] = self.saveData.KillSpree.GoldCoast, -- Gold Coast
        -- Aldmeri Dominion
        ["The Thalmor"] = self.saveData.KillSpree.Auridon, -- Auridon
        ["Back to the"] = self.saveData.KillSpree.Grahtwood, -- Grahtwood
        ["The damn El"] = self.saveData.KillSpree.Greenshade, -- Greenshade
        ["Malabal Tor"] = self.saveData.KillSpree.MalabalTor, -- Malabal Tor
        ["This one do"] = self.saveData.KillSpree.ReapersMarch, -- Reaper's March
        -- Daggerfall Covenant
        ["The Redguar"] = self.saveData.KillSpree.AlikrDesert, -- Alik'r Desert
        ["Strained re"] = self.saveData.KillSpree.Bangkorai, -- Bangkorai
        ["Trade is th"] = self.saveData.KillSpree.Glenumbra, -- Glenumbra
        ["The Covenan"] = self.saveData.KillSpree.Rivenspire, -- Rivenspire
        ["Smuggling i"] = self.saveData.KillSpree.Stormhaven, -- Stormhaven
        -- Ebonheart Pact
        ["Never forgi"] = self.saveData.KillSpree.Deshaan, -- Deshaan
        ["The Thanes "] = self.saveData.KillSpree.Eastmarch, -- Eastmarch
        ["The Ebonhea"] = self.saveData.KillSpree.Shadowfen, -- Shadowfen
        ["Brothers an"] = self.saveData.KillSpree.Stonefalls, -- Stonefalls
        ["My exile le"] = self.saveData.KillSpree.TheRift, -- The Rift
    }
end

--------------------
-- Menu Functions --
--------------------

function RepeatableQuestFilter:CreateSettingsWindow(self)
    local panelData = {
        type = "panel",
        name = "Quest Filter",
        displayName = "Repeatable Quest Filter",
        author = "Positron",
        version = "1.2",
        website = "https://github.com/alexgurrola/RepeatableQuestFilter",
        slashCommand = "/questfilter",
        registerForDefaults = true,
    }
    -- local panel =
    LAM2:RegisterAddonPanel(self.name .. "Config", panelData)
    local optionsData = {}
    optionsData[#optionsData + 1] = {
        type = "header",
        name = "Thieves Guild"
    }
    --[[
    optionsData[#optionsData + 1] = {
        type = "description",
        text = "Enable focused farming or leveling by filtering repeatable quests."
    }
    ]]
    optionsData[#optionsData + 1] = {
        type = "checkbox",
        name = "Filters",
        tooltip = "Turn this off if you want to allow all Thieves Guild Quests through.",
        requiresReload = true,
        default = self.defaults.ThievesGuild,
        getFunc = function()
            return self.saveData.ThievesGuild
        end,
        setFunc = function(newValue)
            self.saveData.ThievesGuild = newValue
        end,
    }
    optionsData[#optionsData + 1] = {
        type = "checkbox",
        name = "Crime Spree",
        tooltip = "Turn this on if you want to allow Thieves Guild Crime Sprees through.",
        requiresReload = true,
        default = self.defaults.CrimeSpree,
        getFunc = function()
            return self.saveData.CrimeSpree
        end,
        setFunc = function(newValue)
            self.saveData.CrimeSpree = newValue
        end,
    }
    optionsData[#optionsData + 1] = {
        type = "checkbox",
        name = "Laundering",
        tooltip = "Turn this on if you want to allow The Covetous Countess through.",
        requiresReload = true,
        default = self.defaults.Launder,
        getFunc = function()
            return self.saveData.Launder
        end,
        setFunc = function(newValue)
            self.saveData.Launder = newValue
        end,
    }
    -- Dark Brotherhood Options
    optionsData[#optionsData + 1] = {
        type = "header",
        name = "Dark Brotherhood"
    }
    optionsData[#optionsData + 1] = {
        type = "checkbox",
        name = "Filters",
        tooltip = "Turn this off if you want to allow all Dark Brotherhood Quests through.",
        requiresReload = true,
        default = self.defaults.DarkBrotherhood,
        getFunc = function()
            return self.saveData.DarkBrotherhood
        end,
        setFunc = function(newValue)
            self.saveData.DarkBrotherhood = newValue
        end,
    }
    -- Neutral
    optionsData[#optionsData + 1] = {
        type = "submenu",
        name = "Neutral Kill Sprees",
        tooltip = "Decide which places you wish to have Kill Sprees available in Neutral Zones.",
        controls = self.controls.KillSpree.Neutral
    }
    self.controls.KillSpree.Neutral[#self.controls.KillSpree.Neutral + 1] = {
        type = "checkbox",
        name = "Gold Coast",
        tooltip = "Turn this on if you want to allow Gold Coast Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.GoldCoast,
        getFunc = function()
            return self.saveData.KillSpree.GoldCoast
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.GoldCoast = newValue
        end,
    }
    -- Aldmeri Dominion
    optionsData[#optionsData + 1] = {
        type = "submenu",
        name = "Aldmeri Dominion Kill Sprees",
        tooltip = "Decide which places you wish to have Kill Sprees available in the Aldmeri Dominion.",
        controls = self.controls.KillSpree.AldmeriDominion
    }
    self.controls.KillSpree.AldmeriDominion[#self.controls.KillSpree.AldmeriDominion + 1] = {
        type = "checkbox",
        name = "Auridon",
        tooltip = "Turn this on if you want to allow Auridon Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Auridon,
        getFunc = function()
            return self.saveData.KillSpree.Auridon
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Auridon = newValue
        end,
    }
    self.controls.KillSpree.AldmeriDominion[#self.controls.KillSpree.AldmeriDominion + 1] = {
        type = "checkbox",
        name = "Grahtwood",
        tooltip = "Turn this on if you want to allow Grahtwood Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Grahtwood,
        getFunc = function()
            return self.saveData.KillSpree.Grahtwood
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Grahtwood = newValue
        end,
    }
    self.controls.KillSpree.AldmeriDominion[#self.controls.KillSpree.AldmeriDominion + 1] = {
        type = "checkbox",
        name = "Greenshade",
        tooltip = "Turn this on if you want to allow Greenshade Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Greenshade,
        getFunc = function()
            return self.saveData.KillSpree.Greenshade
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Greenshade = newValue
        end,
    }
    self.controls.KillSpree.AldmeriDominion[#self.controls.KillSpree.AldmeriDominion + 1] = {
        type = "checkbox",
        name = "Malabal Tor",
        tooltip = "Turn this on if you want to allow Malabal Tor Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.MalabalTor,
        getFunc = function()
            return self.saveData.KillSpree.MalabalTor
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.MalabalTor = newValue
        end,
    }
    self.controls.KillSpree.AldmeriDominion[#self.controls.KillSpree.AldmeriDominion + 1] = {
        type = "checkbox",
        name = "Reaper's March",
        tooltip = "Turn this on if you want to allow Reaper's March Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.ReapersMarch,
        getFunc = function()
            return self.saveData.KillSpree.ReapersMarch
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.ReapersMarch = newValue
        end,
    }
    -- Daggerfall Covenant
    optionsData[#optionsData + 1] = {
        type = "submenu",
        name = "Daggerfall Covenant Kill Sprees",
        tooltip = "Decide which places you wish to have Kill Sprees available in the Daggerfall Covenant.",
        controls = self.controls.KillSpree.DaggerfallCovenant
    }
    self.controls.KillSpree.DaggerfallCovenant[#self.controls.KillSpree.DaggerfallCovenant + 1] = {
        type = "checkbox",
        name = "Alik'r Desert",
        tooltip = "Turn this on if you want to allow Alik'r Desert Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.AlikrDesert,
        getFunc = function()
            return self.saveData.KillSpree.AlikrDesert
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.AlikrDesert = newValue
        end,
    }
    self.controls.KillSpree.DaggerfallCovenant[#self.controls.KillSpree.DaggerfallCovenant + 1] = {
        type = "checkbox",
        name = "Bangkorai",
        tooltip = "Turn this on if you want to allow Bangkorai Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Bangkorai,
        getFunc = function()
            return self.saveData.KillSpree.Bangkorai
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Bangkorai = newValue
        end,
    }
    self.controls.KillSpree.DaggerfallCovenant[#self.controls.KillSpree.DaggerfallCovenant + 1] = {
        type = "checkbox",
        name = "Glenumbra",
        tooltip = "Turn this on if you want to allow Glenumbra Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Glenumbra,
        getFunc = function()
            return self.saveData.KillSpree.Glenumbra
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Glenumbra = newValue
        end,
    }
    self.controls.KillSpree.DaggerfallCovenant[#self.controls.KillSpree.DaggerfallCovenant + 1] = {
        type = "checkbox",
        name = "Rivenspire",
        tooltip = "Turn this on if you want to allow Rivenspire Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Rivenspire,
        getFunc = function()
            return self.saveData.KillSpree.Rivenspire
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Rivenspire = newValue
        end,
    }
    self.controls.KillSpree.DaggerfallCovenant[#self.controls.KillSpree.DaggerfallCovenant + 1] = {
        type = "checkbox",
        name = "Stormhaven",
        tooltip = "Turn this on if you want to allow Stormhaven Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Stormhaven,
        getFunc = function()
            return self.saveData.KillSpree.Stormhaven
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Stormhaven = newValue
        end,
    }
    -- Ebonheart Pact
    optionsData[#optionsData + 1] = {
        type = "submenu",
        name = "Ebonheart Pact Kill Sprees",
        tooltip = "Decide which places you wish to have Kill Sprees available in the Ebonheart Pact.",
        controls = self.controls.KillSpree.EbonheartPact
    }
    self.controls.KillSpree.EbonheartPact[#self.controls.KillSpree.EbonheartPact + 1] = {
        type = "checkbox",
        name = "Deshaan",
        tooltip = "Turn this on if you want to allow Deshaan Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Deshaan,
        getFunc = function()
            return self.saveData.KillSpree.Deshaan
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Deshaan = newValue
        end,
    }
    self.controls.KillSpree.EbonheartPact[#self.controls.KillSpree.EbonheartPact + 1] = {
        type = "checkbox",
        name = "Eastmarch",
        tooltip = "Turn this on if you want to allow Eastmarch Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Eastmarch,
        getFunc = function()
            return self.saveData.KillSpree.Eastmarch
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Eastmarch = newValue
        end,
    }
    self.controls.KillSpree.EbonheartPact[#self.controls.KillSpree.EbonheartPact + 1] = {
        type = "checkbox",
        name = "Shadowfen",
        tooltip = "Turn this on if you want to allow Shadowfen Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Shadowfen,
        getFunc = function()
            return self.saveData.KillSpree.Shadowfen
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Shadowfen = newValue
        end,
    }
    self.controls.KillSpree.EbonheartPact[#self.controls.KillSpree.EbonheartPact + 1] = {
        type = "checkbox",
        name = "Stonefalls",
        tooltip = "Turn this on if you want to allow Stonefalls Kill Sprees through.",
        requiresReload = true,
        default = self.defaults.KillSpree.Stonefalls,
        getFunc = function()
            return self.saveData.KillSpree.Stonefalls
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.Stonefalls = newValue
        end,
    }
    self.controls.KillSpree.EbonheartPact[#self.controls.KillSpree.EbonheartPact + 1] = {
        type = "checkbox",
        name = "The Rift",
        tooltip = "Turn this on if you want to allow Kill Sprees through in The Rift.",
        requiresReload = true,
        default = self.defaults.KillSpree.TheRift,
        getFunc = function()
            return self.saveData.KillSpree.TheRift
        end,
        setFunc = function(newValue)
            self.saveData.KillSpree.TheRift = newValue
        end,
    }
    -- Other Actions
    optionsData[#optionsData + 1] = {
        type = "header",
        name = "Other Actions"
    }
    -- Donation Button
    optionsData[#optionsData + 1] = {
        type = "button",
        name = "Small Donation",
        tooltip = "This will compose appreciation mail for @PositronXX.  You can either accept the prompt directly or decline the transaction and mail a custom letter in the Outbox.",
        func = function()
            local wallet = GetCurrencyAmount(CURT_MONEY, CURRENCY_LOCATION_CHARACTER)
            local tithe = math.floor(wallet / 10)
            MAIL_SEND:ClearFields()
            MAIL_SEND.to:SetText("@PositronXX")
            MAIL_SEND.subject:SetText("Donation to Quest Filter")
            MAIL_SEND.body:SetText("Keep the updates coming!")
            MAIL_SEND:SetSendMoneyMode(true)
            MAIL_SEND:AttachMoney(0, tithe)
            MAIL_SEND:UpdateMoneyAttachment()
            MAIL_SEND:UpdatePostageMoney()
            SCENE_MANAGER:CallWhen("mailSend", SCENE_SHOWN, function()
                ZO_MailSendBodyField:TakeFocus()
            end)
            ZO_MainMenuSceneGroupBar.m_object:SelectDescriptor("mailSend")
            MAIL_SEND:Send()
        end,
    }
    LAM2:RegisterOptionControls(self.name .. "Config", optionsData)
end

------------------------
-- Core Functionality --
------------------------

-- override the chatter option function, so only the filtered quests can be started
function RepeatableQuestFilter:OverwritePopulateChatterOption(self, interaction)
    local _self = self
    local PopulateChatterOption = interaction.PopulateChatterOption
    interaction.PopulateChatterOption = function(self, index, fun, txt, type, ...)
        -- check if the current target is a filtered quest giver
        if not _self.filters.questGiver[lastInteractableName] then
            PopulateChatterOption(self, index, fun, txt, type, ...)
            return
        end
        -- the player has to be on an enabled map
        if not _self.filters.zones[GetZoneId(GetUnitZoneIndex("player"))] then
            PopulateChatterOption(self, index, fun, txt, type, ...)
            return
        end
        -- check if the current dialog starts an enabled quest
        local offerText = GetOfferedQuestInfo()
        if string.len(offerText) == 0 then
            PopulateChatterOption(self, index, fun, txt, type, ...)
            return
        end
        if not _self.filters.dialog[string.sub(offerText, 2, 12)] then
            -- if it is a different quest, only display the goodbye option
            if type ~= CHATTER_GOODBYE then
                return
            end
            PopulateChatterOption(self, 1, fun, txt, type, ...)
            return
        end
        PopulateChatterOption(self, index, fun, txt, type, ...)
        lastInteractableName = nil -- set this variable to nil, so the next dialog step isn't manipulated
    end
end

----------------------
--  Register Events --
----------------------

EVENT_MANAGER:RegisterForEvent(RepeatableQuestFilter.name, EVENT_ADD_ON_LOADED, OnAddOnLoaded)
